### DISCUSSION TITLE
Create a vips virtual image using a tile generator that I provide

### DISCUSSION THREAD

**Discussion by dougwood** [libvips/libvips#4145] [Q&A]
My question is similar to this: https://github.com/libvips/libvips/discussions/2860

I have some very large digital pathology images that are stored in various formats some of which are not natively supported by libvips or even very well by openslide for that matter (Zeiss .czi. for example).  In my image processing chain I would like to run these through a resize() and a mapim() using VIPS, but this of course requires getting the pixels into a vips image object first.  I have a routine to do this which creates the desired VIPS image width, height and pixel type, then reads the czi file tile by tile using the python library provided by Zeiss that reads these files.  It then inserts the memory numpy tiles into a vips image, but this image is memory resident, not virtual.  Unfortunately, this is using tens of GB of system memory which is a resource problem.

Is there some way to create a vips virtual image by providing a tile generator, much like the discussion https://github.com/libvips/libvips/discussions/2860?  
I imagine that this might have many other applications besides mine.

Thank you for your help!

**Comment by jcupitt** [comment]
Hi @dougwood,

You can do this, but you'd need to use C or C++. You'd copy-paste one of the existing loaders (perhaps the webp one?), attach it to your CZI code, and then compile it as a libvips plugin.

There's been talk of making some kind of API that python could use, but unfortunately no one's done the work (as you saw).

**Reply by dougwood** [reply]
Thanks @jcupitt for the reply.  Unfortunately, the library I have for reading czi files is in python. I'm not sure if a C/C++ version has been published. I'll look into that.
